Add ee_alloc_context

noahfalk opened the pull request:
This change is some preparatory refactoring for the randomized allocation sampling feature. We need to add more state onto allocation context but we don't want to do a breaking change of the GC interface. The new state only needs to be visible to the EE but we want it physically near the existing alloc context state for good cache locality. To accomplish this we created a new ee_alloc_context struct which contains an instance of gc_alloc_context within it.

A new field called ee_alloc_context::combined_limit should be used by fast allocation helpers to determine when to go down the slow path. Most of the time combined_limit has the same value as alloc_limit, but periodically we need to emit an allocation sampling event on an object that is somewhere in the middle of an AC. Using combined_limit rather than alloc_limit as the slow path trigger allows us to keep all the sampling event logic in the slow path.

There is another PR for NativeAOT making the same change: https://github.com/dotnet/runtime/pull/104851

noahfalk committed:
Add ee_alloc_context

This change is some preparatory refactoring for the randomized allocation sampling feature. We need to add more state onto allocation context but we don't want to do a breaking change of the GC interface. The new state only needs to be visible to the EE but we want it physically near the existing alloc context state for good cache locality. To accomplish this we created a new ee_alloc_context struct which contains an instance of gc_alloc_context within it.

The new ee_alloc_context.combined_limit field should be used by fast allocation helpers to determine when to go down the slow path. Most of the time combined_limit has the same value as alloc_limit, but periodically we need to emit an allocation sampling event on an object that is somewhere in the middle of an AC. Using combined_limit rather than alloc_limit as the slow path trigger allows us to keep all the sampling event logic in the slow path.

jkotas commented on src/coreclr/vm/gcheaputilities.h:
```diff
@@ -21,7 +60,11 @@ GVAL_DECL(GCHeapType, g_heap_type);
-GVAL_DECL(gc_alloc_context, g_global_alloc_context);
+extern "C" ee_alloc_context g_global_ee_alloc_context;
+
+// This is a pointer into the g_global_ee_alloc_context.gc_allocation_context, the GC visible portion
```
I do not understand why we need this extra indirection via a static variable Why can't all places that use this just take an address of `g_global_alloc_context` field?

noahfalk replied:
Do you mean take the address of g_global_ee_alloc_context.gc_allocation? That would work fine. I didn't do it that way initially just to leave the code a little closer to its current state.

jkotas replied:
Yes, it is what I meant.

noahfalk commented:
This is ready for review now.
All failures are known pre-existing failures according to build analysis. I believe I've applied your feedback as well @jkotas

noahfalk commented:
I converted back to draft because testing on NativeAOT revealed a race condition that I believe effects the CoreCLR work as well. I am investigating.

noahfalk commented on src/coreclr/vm/gcenv.ee.cpp:
```diff
@@ -55,6 +55,27 @@ void GCToEEInterface::RestartEE(bool bFinishedGC)
+    // The GC may change allocation contexts during the GC using GcEnumAllocContexts() so we need to
+    // update the corresponding combined limits now that the GC is complete. Doing this within
+    // GcEnumAllocContexts() is challenging to do correctly or efficiently because multiple GC threads
+    // may enumerate and modify the allocation contexts concurrently.
```
Previously this change was doing the updates inside of GcEnumAllocContexts(). None of the CoreCLR tests were failing but as best I can tell there was a race condition doing it that way.

jkotas replied:
Also, `GcEnumAllocContexts` is enumeration API as its name suggests, so modifying as side-effect would be a plain broken design.

noahfalk replied:
The GC does currently modify ACs inside that enumerator. I think changing that behavior in the GC and modifying the GC<->EE interface is outside the scope of this PR.

jkotas replied:
I do not see the problem with GC side changing the state in the enumerator callback. I was commenting on the VM side enumerator updating state as side-effect of the enumerator.

noahfalk commented:
All issues I know of are resolved and build analysis is green, so this is ready for review once more. Thanks!

jkotas commented on src/coreclr/debug/daccess/dacdbiimpl.cpp:
```diff
@@ -6551,10 +6551,11 @@ HRESULT DacHeapWalker::Init(CORDB_ADDRESS start, CORDB_ADDRESS end)
             }
         }
-        if ((&g_global_alloc_context)->alloc_ptr != nullptr)
+        gc_alloc_context globalCtx = ((ee_alloc_context)g_global_alloc_context).gc_allocation_context;
```
Why do we need the `ee_alloc_context` cast here? 

Also, creating a full copy of the `ee_alloc_context` on stack looks unnecessary - use pointer to `ee_alloc_context` instead?

noahfalk replied:
> Why do we need the ee_alloc_context cast here?

g_global_alloc_context is of type __GlobalVal<ee_alloc_context> so dotting directly without a cast wouldn't work.

> use pointer to ee_alloc_context instead?

You could do this:
```
gc_alloc_context* pGlobalCtx = &(&g_global_alloc_context)->gc_allocation_context;
```
Or this:
```
gc_alloc_context* pGlobalCtx = PTR_TO_MEMBER_TADDR(ee_alloc_context, &g_global_alloc_context, gc_allocation_context);
```

Personally, I thought neither of those were as readable and I expect the perf cost in every case is dominated by issuing a remote memory read.

noahfalk replied:
Unless I hear that is more important than I realize I'm planning to leave this code as-is.

jkotas replied:
I guess it is ok if it works. It just does not look like typical DACized code.

jkotas commented on src/coreclr/debug/daccess/request.cpp:
```diff
@@ -5493,8 +5493,9 @@ HRESULT ClrDataAccess::GetGlobalAllocationContext(
     SOSDacEnter();
-    *allocPtr = (CLRDATA_ADDRESS)((&g_global_alloc_context)->alloc_ptr);
-    *allocLimit = (CLRDATA_ADDRESS)((&g_global_alloc_context)->alloc_limit);
+    gc_alloc_context global_alloc_context = ((ee_alloc_context)g_global_alloc_context).gc_allocation_context;
```
Dtto

noahfalk replied:
Unless I hear that is more important than I realize I'm planning to leave this code as-is (see discussion above)

jkotas commented on src/native/managed/cdacreader/src/DataType.cs:
```diff
@@ -23,6 +23,7 @@ public enum DataType
     Thread,
     ThreadStore,
     GCAllocContext,
+    EEAllocContext,
```
@elinor-fung @lambdageek Please review the contract additions

noahfalk replied:
ping @elinor-fung @lambdageek - any concerns here? This is the only outstanding part of the code review remaining.

jkotas commented on src/coreclr/vm/gcenv.ee.cpp:
```diff
@@ -55,6 +55,27 @@ void GCToEEInterface::RestartEE(bool bFinishedGC)
 {
     WRAPPER_NO_CONTRACT;
 
+    // The GC may change allocation contexts during the GC using GcEnumAllocContexts() so we need to
```
It is unfortunate that we have do a dedicated thread enumeration of the thread list just to clear the single field. Is there really no better way to do that?

cc @VSadov

VSadov replied:
This can be noticeable on scenarios like asp.net benchmarks (relatively small heap, but lots of threads), but may not be a big deal overall next to the cost of a GC.

Perhaps should check that suspend reason was indeed `SUSPEND_FOR_GC`. 
I'd assume that other kinds of suspension like `SUSPEND_FOR_GC_PREP` will not be adjusting allocation budgets.

noahfalk replied:
> Is there really no better way to do that?

I chatted with Maoni and she was comfortable with us codifying as part of the current GC contract that the only legal modification the GC can make inside the enumeration is to set ptr = limit = 0. Once the EE is allowed to assume that additional information then we can change the enumeration code to look like this:

```
fn(alloc_context, param);
if(alloc_context->alloc_limit == 0)
{
    combined_limit = 0;
}
```

Concurrent threads enumerating will still create races, but the races will no longer matter. How does that sound?

noahfalk replied:
I just noticed your comment https://github.com/dotnet/runtime/pull/104849#discussion_r1677867717 @jkotas. We can update combined_limit inline inside GCEnumAllocContexts() or we can do it with a separate loop through all the threads afterwards.

noahfalk replied:
I've updated the PR with the new approach that zeros combined_limit inside the GCEnumAllocContexts() function and avoids doing a separate thread enumeration in RestartEE().

jkotas replied:
> Concurrent threads enumerating will still create races,

It is also going to create cache contention.

It would be much easier to reason about an explicit contract where the GC can say that it is doing the final fixup of the allocation context.

noahfalk replied:
The plan for now is to do some perf testing with a large number of threads but @Maoni0 did not expect this to be an issue. She said that past GC performance investigations had not raised these thread enumerations playing a significant role.

Fwiw, the GC's existing code already creates similar contention here. N GC server threads run the thread enumeration in parallel and the callback both reads and potentially writes to fields that reside in the same cache line. https://github.com/dotnet/runtime/blob/main/src/coreclr/gc/gc.cpp#L7888-L7896. The likely case is that whatever thread dirties the cache line writing alloc_ptr=0 and alloc_limit=0 will be the same thread that writes combined_limit=0 a few CPU instructions later.

noahfalk replied:
@jkotas - For now the options I am aware of are either the RestartEE() separate thread enumeration or the inline updates on GcEnumAllocContexts(). I am not aware of any other options that don't require changes to the GC<->EE interface and @Maoni0 doesn't have time available to discuss/design/approve such changes. You've raised criticism on both choices, but I think the path forward requires either picking one of them or suggesting a new 3rd option which doesn't require GC interface changes to implement.

jkotas replied:
I think we should be open to adding new methods to GC<->EE interface as necessary. GC<->EE interface was not designed as internal implementation details that is not set in stone, it is very similar to JIT<->EE interface. Changes in JIT<->EE interface is no-brainer that allows us to evolve the system as something that easy to reason about and efficient. I do not understand why changing the GC<->EE interface is such a big deal.

I think that the design of the new GC<->EE interface method should be pretty straightforward: Add new variant of `GcEnumAllocContexts` that the GC is going to use when it is repairing the contexts to let the EE side know that it needs to do its part. Am I missing something?

noahfalk replied:
We are effectively at 'disagree and commit'. @Maoni0 does not wish to invest in any GC<->EE interface changes right now as a matter of prioritization. @jkotas notes his disagreement, but we'll move forward with no GC<->EE change and the PR will stay as is. I will confirm its not causing a perf issue.

jkotas replied:
I think that this change is regressing the GC/EE interface design. As discussed offline, I can live with it but I am not happy about it.

noahfalk committed:
PR feedback

combined_limit is now synchronized in GcEnumAllocContexts instead of RestartEE.
This requires the GC being constrained in how it updates the alloc_ptr and alloc_limit. No GC behavior changed,
in practice, but the constraints are now part of the EE<->GC contract so that we can rely on them in the EE code.

jkotas commented on src/coreclr/vm/gcheaputilities.h:
```diff
@@ -12,6 +12,69 @@ GPTR_DECL(IGCHeap, g_pGCHeap);
+    void init()
+    {
+        LIMITED_METHOD_CONTRACT;
+        combined_limit = 0;
```
```suggestion
        combined_limit = NULL;
```
`combined_limit` is a pointer. NULL or nullptr is more appropriate. Some compilers tend to be picky about 0 vs. NULL.

jkotas replied:
Hmm, I see we `gc_alloc_context` uses 0 currently. I guess the compilers are not complaining about this case yet. You can ignore this comment.

jkotas commented on src/coreclr/vm/gcenv.ee.cpp:
```diff
@@ -460,16 +460,29 @@ void GCToEEInterface::GcEnumAllocContexts(enum_alloc_context_func* fn, void* par
+                // combined_limit up-to-date. Note that the GC has multiple threads running this enumeration concurrently
+                // with no synchronization. If you need to change this code think carefully about how that concurrency
+                // may affect the results.
+                if(ac->alloc_limit == 0)
```
```suggestion
                if (ac->alloc_limit == 0)
```

jkotas commented on src/coreclr/vm/gcenv.ee.cpp:
```diff
@@ -460,16 +460,29 @@ void GCToEEInterface::GcEnumAllocContexts(enum_alloc_context_func* fn, void* par
     {
-        fn(&g_global_alloc_context, param);
+        fn(&g_global_alloc_context.gc_allocation_context, param);
+        if(g_global_alloc_context.gc_allocation_context.alloc_limit == 0)
```
```suggestion
        if (g_global_alloc_context.gc_allocation_context.alloc_limit == 0)
```

jkotas commented:
Is it that hard to implement it in a way that it does not introduce cache contention from thread enumerations running in parallel? I think it is preferable to go the designs that do not have performance issues by construction.

noahfalk committed:
Apply suggestions from code review

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

jkotas commented on src/coreclr/vm/gcenv.ee.cpp:
```diff
@@ -460,16 +460,29 @@ void GCToEEInterface::GcEnumAllocContexts(enum_alloc_context_func* fn, void* par
+                // combined_limit up-to-date. Note that the GC has multiple threads running this enumeration concurrently
+                // with no synchronization. If you need to change this code think carefully about how that concurrency
+                // may affect the results.
+                if (ac->alloc_limit == 0)
```
```suggestion
                if (ac->alloc_limit == 0 && ac->combined_limit != 0)
```
This should reduce the cache contention when multiple GC threads are enumerating the threads

noahfalk replied:
sure thing. I expect to be running some perf tests later today and I'll make sure this is included.

jkotas commented on src/coreclr/vm/gcenv.ee.cpp:
```diff
@@ -460,16 +460,29 @@ void GCToEEInterface::GcEnumAllocContexts(enum_alloc_context_func* fn, void* par
     {
-        fn(&g_global_alloc_context, param);
+        fn(&g_global_alloc_context.gc_allocation_context, param);
+        if (g_global_alloc_context.gc_allocation_context.alloc_limit == 0)
```
```suggestion
        if (g_global_alloc_context.gc_allocation_context.alloc_limit == 0 && g_global_alloc_context.gc_allocation_context.alloc_limit != 0)
```
For symmetry, though it won't make much difference for UP systems.

noahfalk commented:
@jkotas Following up on perf, I attempted to stress the threading by running gcperfsim with 400 threads and gc server config, allocating 500GB of objects. I used two different configs, one with a 50MB heap size and small objects and the other with a 2gb heap size allowing for a 20% LOH allocation mix to introduce more BGCs. My machine has 20 hardware threads. Median execution time in scenario 1 was 34.5 sec for PR and baseline. Scenario 2 was 38.4 sec for baseline and 38.5 sec for the PR. Noise on machine causes runs to vary by up to 1 second so the 0.1 diff isn't statistically significant. It's possible with a large sample size and a low noise environment we could discern some cost from the noise, but it certainly doesnt jump out. I also took CPU sampling traces which showed GCEnumAllocContexts consistently consumed 0.3% inclusive time before and after the change.

Also fyi I lost internet service at my house. I'm trying to get it restored but for now I only have GH access from my cell phone.

noahfalk committed:
Apply suggestions from code review

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

noahfalk commented:
@jkotas - anything remaining before you can sign off? I believe everything has been resolved aside from still waiting to hear from @elinor-fung or @lambdageek.

noahfalk committed:
Fix dac-ization bug

The code of GetAllocContext() was constructing a PTR_gc_alloc_context which does a host->target pointer conversion. Those conversions work by doing a lookup in a dictionary of blocks of memory that we have previously marshalled and the pointer being converted is expected to be the start of the memory block. In this case we had never previously marshalled the gc_allocation_context on its own. We had only marshalled the m_pRuntimeThreadLocals block which includes the gc_allocation_context inside of it at a non-zero offset. This caused the host->target pointer conversion to fail which in turn meant commands like !threads in SOS would fail.

The fix is pretty trivial. We don't need to do a host->target conversion here at all because the calling code in the DAC is going to immediately convert right back to a host pointer. We can avoid the conversion in both directions by eliminating the cast and returning the host pointer directly.

noahfalk commented:
I did some manual testing of the debugging scenario. In the normal DAC scenario I found and fixed one bug in the new commit. CDAC appeared to be working correctly. My test was running windbg + SOS with !threads and !DumpHeap to exercise reading gc_alloc_context. I confirmed using a debugger on the debugger that the CDAC case was executing down the CDAC branch of the DAC code as intended.

jkotas commented:
> perf

To assess perf impact for changes like this, we typically try to craft a microbenchmark that tries to show worst regression or best improvement (one example: https://github.com/dotnet/runtime/pull/96150#issuecomment-1863434492). Unless you introduce very bad bug, the impact will be lost in the noise from all other things going on in a general-purpose allocation throughput microbenchmark.

I am not worried about the perf impact too much with the latest version of the change, so I do not think it is necessary to do that.

jkotas commented:
/azp run runtime-coreclr gcstress-extra

jkotas commented:
GC stress crash at:
```
      Assert failure(PID 2208 [0x000008a0], Thread: 9372 [0x249c]): Consistency check failed: AV in clr at this callstack:
      ------
      CORECLR! Thread::CooperativeCleanup + 0xFF (0x6fff547e)
      CORECLR! Thread::DetachThread + 0x124 (0x6fff69ec)
      CORECLR! TlsDestructionMonitor::~TlsDestructionMonitor + 0x8C (0x7043d0d8)
      CORECLR! _dyn_tls_dtor + 0x8A (0x7045116a)
```
It looks like a regression introduced by this change (`Thread::CooperativeCleanup` is modified by this change).

noahfalk committed:
Fix test failure

Somehow a test reached Thread::CooperativeCleanup() with m_pRuntimeThreadLocals==NULL. Looking at the code I expected that would mean GetThread()==NULL or ThreadState contains TS_Dead, but neither of those conditions were true so it is unclear how it executed to that state. The callstack was:
>	coreclr.dll!Thread::CooperativeCleanup() Line 2762	C++
 	coreclr.dll!Thread::DetachThread(int fDLLThreadDetach) Line 936	C++
 	coreclr.dll!TlsDestructionMonitor::~TlsDestructionMonitor() Line 1745	C++
 	coreclr.dll!__dyn_tls_dtor(void * __formal, const unsigned long dwReason, void * __formal) Line 122	C++
 	ntdll.dll!_LdrxCallInitRoutine@16()	Unknown
 	ntdll.dll!LdrpCallInitRoutine()	Unknown
 	ntdll.dll!LdrpCallTlsInitializers()	Unknown
 	ntdll.dll!LdrShutdownThread()	Unknown
 	ntdll.dll!RtlExitUserThread()	Unknown

Regardless, the previous code wouldn't have hit that issue because it obtained the pointer through t_runtime_thread_locals rather than m_pRuntimeThreadLocals. I restored using t_runtime_thread_locals in the Cleanup routine. Out of caution I also searched for any other places that were previously accessing the alloc_context through the thread local address and ensured they don't switch to use m_pRuntimeThreadLocals either.

noahfalk committed:
Code review feedback - Renaming fields using m_ style

noahfalk commented:
I've updated for the feedback so far + the test failure that showed up in the GC stress run. Looking at the time I think I may need to call it quits on this for .NET 9 and instead treat it as .NET 10 on a slower cadence. At this point I don't think I have more time available right now to do a staged checkin, revalidate testing after the ongoing changes, and continue to have time in reserve to respond to potential post-checkin issues.

noahfalk commented:
I'm resuming work on this feature. @jkotas as best I'm aware all the comments you had made on the PR had been addressed and the GC stress issue was already fixed back in July. Let me know if there is anything else before we get this merged. Thanks!
